packages/react-native/ReactCommon/react/performance/timeline/PerformanceEntryReporter.cpp

```diff
@@ -17,199 +17,69 @@ PerformanceEntryReporter::getInstance() {
-  std::stable_sort(
-      res.entries.begin(),
-      res.entries.end(),
-      [](const PerformanceEntry& lhs, const PerformanceEntry& rhs) {
-        if (lhs.startTime != rhs.startTime) {
-          return lhs.startTime < rhs.startTime;
-        } else {
-          return lhs.duration < rhs.duration;
-        }
-      });
-
-  droppedEntriesCount_ = 0;
-  return res;
-}
-
-void PerformanceEntryReporter::logEntry(const PerformanceEntry& entry) {
-  if (entry.entryType == PerformanceEntryType::EVENT) {
-    eventCounts_[entry.name]++;
-  }
-
-  if (!isReporting(entry.entryType) && !isAlwaysLogged(entry.entryType)) {
-    return;
-  }
-
-  std::lock_guard lock(entriesMutex_);
-
-  auto& buffer = getBuffer(entry.entryType);
-
-  if (entry.duration < buffer.durationThreshold) {
-    // The entries duration is lower than the desired reporting threshold, skip
-    return;
-  }
-
-  if (buffer.hasNameLookup) {
-    // If we need to remove an entry because the buffer is null,
-    // we also need to remove it from the name lookup.
-    auto overwriteCandidate = buffer.entries.getNextOverwriteCandidate();
-    if (overwriteCandidate != nullptr) {
-      std::lock_guard lock2(nameLookupMutex_);
-      auto it = buffer.nameLookup.find(overwriteCandidate);
-      if (it != buffer.nameLookup.end() && *it == overwriteCandidate) {
-        buffer.nameLookup.erase(it);
-      }
-    }
-  }
-
-  auto pushResult = buffer.entries.add(std::move(entry));
-  if (pushResult ==
-      BoundedConsumableBuffer<PerformanceEntry>::PushStatus::DROP) {
-    // Start dropping entries once reached maximum buffer size.
-    // The number of dropped entries will be reported back to the corresponding
-    // PerformanceObserver callback.
-    droppedEntriesCount_ += 1;
-  }
-
-  if (buffer.hasNameLookup) {
-    std::lock_guard lock2(nameLookupMutex_);
-    auto currentEntry = &buffer.entries.back();
-    auto it = buffer.nameLookup.find(currentEntry);
-    if (it != buffer.nameLookup.end()) {
-      buffer.nameLookup.erase(it);
-    }
-    buffer.nameLookup.insert(currentEntry);
-  }
-
-  if (buffer.entries.getNumToConsume() == 1) {
-    // If the buffer was empty, it signals that JS side just has possibly
-    // consumed it and is ready to get more
-    scheduleFlushBuffer();
-  }
+double PerformanceEntryReporter::getDroppedEntriesCount(
+    PerformanceEntryType type) const noexcept {
+  return getBuffer(type).droppedEntriesCount;
 }
 
 void PerformanceEntryReporter::mark(
     const std::string& name,
     const std::optional<DOMHighResTimeStamp>& startTime) {
-  logEntry(PerformanceEntry{
+  const auto entry = PerformanceEntry{
       .name = name,
       .entryType = PerformanceEntryType::MARK,
-      .startTime = startTime ? *startTime : getCurrentTimeStamp()});
+      .startTime = startTime ? *startTime : getCurrentTimeStamp()};
+
+  {
+    std::lock_guard lock(buffersMutex_);
+    markBuffer_.add(entry);
+  }
+
+  observerRegistry_->queuePerformanceEntry(entry);
```

Comment:
In a follow up PR, we should make `mark` and `measure` return the entry so we can pass it back to JS. One problem with the current implementation is that this doesn't work:

```
performance.mark('start-foo');
const entry = performance.measure('foo', 'start-foo');
entry.startTime; // number
entry.duration; // number
```